### ISSUE TITLE
"illegal hardware instruction" On macOS / ARM64

### ISSUE THREAD

**Issue by StanislavNikolov** [odin-lang/Odin#1615]
## Context

Odin fails to compile / run the following simple code on M1 Macbook air with macOS Monterey.

```
package main

import "core:fmt"
import "core:math/rand"

main :: proc() {
	r := rand.create(123);
	number := rand.float32(&r);
	fmt.println(number);
}
```
```
$ odin run wtf_clean.odin
[ no output ]
$ echo $? 
4
$ odin build wtf_clean.odin && ./wtf_clean             
[1]    13178 illegal hardware instruction  ./wtf_clean
$ odin build wtf_clean.odin -microarch:generic && ./wtf_clean 
[1]    13219 illegal hardware instruction  ./wtf_clean
```

* `odin report` output:
 	Odin: dev-2022-03:e6d3e89
	OS:   macOS 12.0.3 Monterey
	CPU:  ARM64
	RAM:  16384 MiB

## Expected Behavior
The compiler should produce an error in case the source code is invalid or the program should run successfully.

## Current Behavior
The compiler produces no error but the program fails to run.

More examples that show this is a bug:
```
$ cat wtf2.odin
package main
import "core:fmt"
import "core:math/rand"
main :: proc() {
	fmt.println("This is the first line of the main");
	r := rand.create(123);
	number := rand.float32(&r);
	fmt.println(number);
}

$ odin build wtf2.odin && ./wtf2
[1]    13515 illegal hardware instruction  ./wtf2
```
If the problem was with my usage of the rand package, I would expect the program to crash after the first println.

```
$ cat wtf3.odin 
package main
import "core:fmt"
import "core:math/rand"
main :: proc() {
	fmt.println("This is the first line of the main");
	// r := rand.create(123);
	// number := rand.float32(&r);
	number := 0;
	fmt.println(number);
}

$ odin build wtf3.odin && ./wtf3
This is the first line of the main
0
```
Removing the usage of the rand.create and rand.float32 fixes the issue. Any of them being present triggers the issue.

**Comment by StanislavNikolov** [comment]
It looks like the issue is not completely fixed, but my proof is quite elaborate.

Look at the following code samples:
## A
```odin
package main

import "core:fmt"
import "core:log"
import "core:math/rand"

import "vendor:sdl2"
CTX :: struct { window: ^sdl2.Window, }
ctx := CTX{}

main :: proc() {
	context.logger = log.create_console_logger()

	fmt.println("BEGIN")
	fmt.println(int(rand.int31_max(1000)))
}
```

## B (same as A, 2 lines commented out)
```odin
package main

import "core:fmt"
// import "core:log"
import "core:math/rand"

import "vendor:sdl2"
CTX :: struct { window: ^sdl2.Window, }
ctx := CTX{}

main :: proc() {
	// context.logger = log.create_console_logger()

	fmt.println("BEGIN")
	fmt.println(int(rand.int31_max(1000)))
}
```

## C (same as A, 3 lines commented out)
```odin
package main

import "core:fmt"
import "core:log"
import "core:math/rand"

// import "vendor:sdl2"
// CTX :: struct { window: ^sdl2.Window, }
// ctx := CTX{}

main :: proc() {
	context.logger = log.create_console_logger()

	fmt.println("BEGIN")
	fmt.println(int(rand.int31_max(1000)))
}
```

A, B and C compile successfully with -opt:3, but only B and C manage to execute successfully. Here's what happens with A
```
$ odin build A.odin -extra-linker-flags:-L/opt/homebrew/lib -vet -opt:3 
$ lldb ./A   
(lldb) target create "./A"
Current executable set to '/Users/stjo/projects/toys/road/A' (arm64).
(lldb) run
Process 58316 launched: '/Users/stjo/projects/toys/road/A' (arm64)
Process 58316 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x46f8de3bf53)
    frame #0: 0x0000000100005e64 A`__$startup_runtime + 12748
A`__$startup_runtime:
->  0x100005e64 <+12748>: stp    x10, x8, [x9, #0x48]
    0x100005e68 <+12752>: adr    x8, #0x32d38              ; rand.global_rand-1131
    0x100005e6c <+12756>: nop    
    0x100005e70 <+12760>: stur   q6, [x1, #0xf8]
Target 0: (A) stopped.
(lldb) thread backtrace
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x46f8de3bf53)
  * frame #0: 0x0000000100005e64 A`__$startup_runtime + 12748
    frame #1: 0x0000000100029950 A`main + 208
    frame #2: 0x00000001000450f4 dyld`start + 520
```

If `core/math/rand.odin` the line
```odin
global_rand := create(u64(intrinsics.read_cycle_counter()))
```
is replaced with
```odin
global_rand := create(u64(0))
```
then all the programs above execute correctly.

TL;DR:
`intrinsics.read_cycle_counter()` + `vendor:sdl2` + `context.logger` + `-opt:3` make program go boom, but remove *any* of them and everything starts working.
